src/utils/sort.hpp

```diff
@@ -0,0 +1,73 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_SORT_HPP_
+#define UTILS_SORT_HPP_
+
+//! \file sort.hpp
+//  \brief Contains functions for sorting data according to a provided comparator
+//  See tst/unit/test_unit_sort.cpp for example usage.
+
+#include "defs.hpp"
+#include "parthenon_arrays.hpp"
+
+#ifdef KOKKOS_ENABLE_CUDA
+#include <thrust/device_ptr.h>
+#include <thrust/sort.h>
+#endif
+
+namespace parthenon {
+
+template <class Key, class KeyComparator>
+void sort(ParArray1D<Key> data, KeyComparator comparator, size_t min_idx,
+          size_t max_idx) {
+  PARTHENON_DEBUG_REQUIRE(min_idx >= 0 && min_idx < data.extent(0),
+                          "Invalid minimum sort index!");
+  PARTHENON_DEBUG_REQUIRE(max_idx >= 0 && max_idx < data.extent(0),
+                          "Invalid maximum sort index!");
+#ifdef KOKKOS_ENABLE_CUDA
+  thrust::device_ptr<Key> first_d = thrust::device_pointer_cast(data.data()) + min_idx;
+  thrust::device_ptr<Key> last_d = thrust::device_pointer_cast(data.data()) + max_idx + 1;
+  thrust::sort(first_d, last_d, comparator);
```

Comment:
As far as I know we don't use `thrust` anywhere else in parthenon, but my understanding is this should always be available when CUDA is available. Still, calling this out explicitly in case anyone has concerns.

Reply:
What happens on Intel and AMD GPUs? The `std::sort` probably won't work without additional magic, does it?
What's the issue with Kokkos sort? You mentioned performance but my quick search in the Kokkos wiki didn't bring up any helpful information.

Reply:
I think @brryan was scared off by this issue:
https://github.com/kokkos/kokkos/issues/2914

For intel and AMD GPUs I suggest we fall back to `Kokkos::sort`. But I like using `thrust::sort` and `std::sort` on appropriate hardware for now, until `Kokkos::sort` is more performant.

Reply:
How about asking the Kokkos team about the current status and doing some (limited) performance tests covering the intended (immediate) use case  ourselves?
My main motivation is to keep the amount of different code paths as low as possible (especially given that the other architecture will become more and more available next year).

Reply:
As I just replied to Jonah, it [sounds like](https://github.com/kokkos/kokkos/issues/4487#issuecomment-963540068) the Kokkos team is looking to rewrite their sorting methods, which should both allow for key-value pair sorting like I'm using here and hopefully improve performance. OK if I leave this PR as-is and open an issue to plan to move to `Kokkos::sort` exclusively when it is ready?

Reply:
To be clear this PR currently will not work for Intel or AMD GPUs as is, but is that a broad requirement now?

Reply:
Following the discussion on the `Kokkos::sort` status, I'm fine with keeping thrust for Cuda for now.
From a functional point of view it'd still be great to call `Kokkos::sort()` if neither `std::sort` nor `thrust` is availabe, or throw a meaningful error message (at compile time).
AMD GPUs may be there sooner than later and I'm currently fixing a couple of things to make those work in #629 as AMD GPUs will become very relevant for us this year.

Reply:
Good point, yes, I should include an error message (done). This might still cause a compiler warning unless the `if(std::is_same<...` is a `constexpr if`, but I assume it will be if compiling in C++17 mode?

Searching real quick, it seems to me that AMD GPUs support their [Bolt library, which has its own sort function?](https://hsa-libraries.github.io/Bolt/html/) However, I can't immediately tell whether that is guaranteed to be included in an AMD compilation environment like thrust is with CUDA. However, if we need a performant sort on AMD in the near term, that is one possibility